### ISSUE TITLE
Update request: jextract 2023-11-27 (JDK 21) → 22-jextract+3-13 (2024/2/21)

### ISSUE THREAD

**Issue by msgilligan** [NixOS/nixpkgs#293102]
- Package name: jextract (jextract-unstable)
- Current version on the unstable channel: 2023-11-27
- Current version on the stable/release channel: n/a

- [x] Checked the [nixpkgs pull requests](https://github.com/NixOS/nixpkgs/pulls)

As of today the current release build is `22-jextract+3-13 (2024/2/21)`  (see https://jdk.java.net/jextract/) It would be nice for the nixpkgs version to be based on JDK 22 -- which contains the final non-preview release of Java Foreign Function & Memory API (aka "Panama")

There are significant changes to `jextract` that users will want before building/launching projects that use Panama. See: https://cr.openjdk.org/~mcimadamore/panama/jextract_changes.html

I am wiling to contribute but before we can do this we'll need to build with JDK 22 (see Issue #271971)

@SharzyL @SpriteOvO

**Comment by SpriteOvO** [comment]
It would be great if we could update `jextract` to be based on JDK 22 :+1: , and we should also keep the current JDK 21 `jextract` package as a new name (perhaps `jextract-21`?) at the same time, as some downstreams still dependent on it and JDK 21 is a LTS version.

**Comment by msgilligan** [comment]
So should the first step be to rename the existing package?  And leave `jextract` as JDK 21-based (with a warning, maybe) and then do a separate PR to upgrade `jextract` to JDK 21?

**Comment by msgilligan** [comment]
Update: There is a PR for JDK 22: https://github.com/NixOS/nixpkgs/pull/295440

**Comment by jlesquembre** [comment]
I can open a PR to update jextract after #295440 gets merged.

@SpriteOvO Considering the significant API changes between JDK 21 and 22, and the fact that jextract itself produces very differnt code if you have to use JDK21 (see https://cr.openjdk.org/~mcimadamore/panama/jextract_changes.html) , is there really value in keeping a jextract version compatible with JDK21?

**Comment by msgilligan** [comment]
> I can open a PR to update jextract after #295440 gets merged.

Awesome! I'm going to test #295440 on macOS/aarch64 this weekend, so that will help get it merged.

> @SpriteOvO Considering the significant API changes between JDK 21 and 22, and the fact that jextract itself produces very differnt code if you have to use JDK21 (see https://cr.openjdk.org/~mcimadamore/panama/jextract_changes.html) , is there really value in keeping a jextract version compatible with JDK21?

@SpriteOvO wrote:
>> as some downstreams still depend on it and JDK 21 is a LTS version.

JDK 21 is LTS, but Panama (Foreign Functions & Memory API) is a "preview feature" which makes it really hard to build anything real with it (i.e. publish JARs for use by other projects. I even had problems with a Gradle multi-module build). I migrated my project to 22 a few weeks ago and am not looking back (which meant I had to stop using Nix as a source for `jextract` temporarily.)

But, I don't see any harm in renaming the existing `jextract` to `jextract-21` and starting a new one. And we certainly don't want to unnecessarily break anyone's dependencies.

**Comment by SpriteOvO** [comment]
> I can open a PR to update jextract after #295440 gets merged.

Nice to hear that, and thanks for your work!

> is there really value in keeping a jextract version compatible with JDK21?

IMO, I think it's worth. The point is that JDK 21 is a LTS version, some downstreams prefer to use a LTS version of JDK to develop their projects. For them, if we don't keep JDK 21's `jextract`, this will be unavailable to them unless they give up on choosing LTS and do some extra effort - porting their project to a newer version of JDK.

**Comment by msgilligan** [comment]
> some downstreams prefer to use a LTS version of JDK to develop their projects. For them, [we should] keep JDK 21's jextract

I agree.

But I also strongly encourage new projects to start with JDK 22 and existing projects to migrate as quickly as possible. Having to make sure "preview mode" is set in the compiler and at runtime can cause various problems with tools and the output of the latest `jextract` is significantly improved. In general, it's much better to be using a non-LTS JDK with a non-preview feature than and LTS JDK with a preview feature.

I migrated my project recently and wish I had migrated toe JDK 22 much earlier.

**Comment by msgilligan** [comment]
PR #295440 has been merged. I gave it a quick test on macOS/aarch64 and it seems to be working. (I used it to build https://github.com/bitcoinj/secp256k1-jdk which uses JDK 22 and Panama via Gradle's Java toolchains)
